File: day-13/python/main.py
import pathlib
import itertools


def find_earliest_bus(
    earliest_timestamp: int, bus_id_indices: list[tuple[int, int]]
) -> int:
    for timestamp in itertools.count(start=earliest_timestamp):
        for _, bus_id in bus_id_indices:
            if timestamp % bus_id == 0:
                return bus_id * (timestamp - earliest_timestamp)


def find_earliest_valid_timestamp(bus_id_indices: list[tuple[int, int]]) -> int:
    earliest_valid_timestamp = step = 1
    for index, bus_id in bus_id_indices:
        earliest_valid_timestamp = next(
            timestamp
            for timestamp in itertools.count(earliest_valid_timestamp, step)
            if (timestamp + index) % bus_id == 0
        )
        step *= bus_id
    return earliest_valid_timestamp


with open(pathlib.Path(__file__).parent.parent / 'input.txt') as f:
    lines = f.read().splitlines()
    bus_id_indices = [
        (index, int(bus_id))
        for index, bus_id in enumerate(lines[1].split(','))
        if bus_id != 'x'
    ]

part_one_solution = find_earliest_bus(int(lines[0]), bus_id_indices)
part_two_solution = find_earliest_valid_timestamp(bus_id_indices)

print('Part One:', part_one_solution)
print('Part Two:', part_two_solution)

assert part_one_solution == 1835
assert part_two_solution == 247086664214628

File: day-13/rust/src/main.rs
use std::fs;
use std::path::Path;

fn find_earliest_bus(earliest_timestamp: usize, bus_id_indices: &[(usize, usize)]) -> usize {
    for timestamp in earliest_timestamp.. {
        for (_, bus_id) in bus_id_indices {
            if timestamp % bus_id == 0 {
                return bus_id * (timestamp - earliest_timestamp);
            }
        }
    }
    0
}

fn find_earliest_valid_timestamp(bus_id_indices: &[(usize, usize)]) -> usize {
    let (mut earliest_valid_timestamp, mut step) = (1, 1);
    for (index, bus_id) in bus_id_indices {
        earliest_valid_timestamp = (earliest_valid_timestamp..)
            .step_by(step)
            .find(|timestamp| (timestamp + index) % bus_id == 0)
            .unwrap();
        step *= bus_id
    }
    earliest_valid_timestamp
}

fn main() {
    let input: Vec<String> = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("input.txt"),
    )
    .unwrap()
    .lines()
    .map(|l| l.parse().unwrap())
    .collect();

    let earliest_timestamp = input[0].parse().unwrap();
    let bus_indices: Vec<_> = input[1]
        .split(',')
        .enumerate()
        .filter(|(_, bus_id)| *bus_id != "x")
        .map(|(index, bus_id)| (index, bus_id.parse().unwrap()))
        .collect();

    let part_one_solution = find_earliest_bus(earliest_timestamp, &bus_indices);
    let part_two_solution = find_earliest_valid_timestamp(&bus_indices);

    println!("Part One: {}", part_one_solution);
    println!("Part Two: {}", part_two_solution);

    assert_eq!(part_one_solution, 1835);
    assert_eq!(part_two_solution, 247086664214628);
}
